packages/library-sync/src/commonMain/kotlin/io/realm/mongodb/internal/SyncSessionImpl.kt

```diff
@@ -15,10 +15,103 @@
  */
 
 package io.realm.mongodb.internal
-
+import io.realm.internal.RealmImpl
+import io.realm.internal.interop.RealmInterop
 import io.realm.internal.interop.RealmSyncSessionPointer
+import io.realm.internal.interop.SyncSessionTransferCompletionCallback
+import io.realm.internal.platform.freeze
+import io.realm.mongodb.SyncErrorCode
+import io.realm.mongodb.SyncException
 import io.realm.mongodb.SyncSession
+import kotlinx.coroutines.TimeoutCancellationException
+import kotlinx.coroutines.channels.Channel
+import kotlinx.coroutines.withContext
+import kotlinx.coroutines.withTimeout
+import kotlin.time.Duration
 
 internal open class SyncSessionImpl(
+    private val realm: RealmImpl?,
     internal val nativePointer: RealmSyncSessionPointer
-) : SyncSession
+) : SyncSession {
+
+    // Constructor used when there is no Realm available, e.g. in the SyncSessionErrorHandler.
+    // Without a Realm reference, it is impossible to track shared state between the public
+    // Realm and the SyncSession. This impacts `downloadAllServerChanges()`.
+    // Since there probably isn't a use case where you ever is going to call
+    // `downloadAllServerChanges` inside the erorr handler, we are just going to disallow it by
+    // throwing an IllegalStateException. Mostly because that is by far the easiest with the
+    // current implementation.
+    constructor(ptr: RealmSyncSessionPointer) : this(null, ptr)
+
+    private enum class TransferDirection {
+        UPLOAD, DOWNLOAD
+    }
+
+    override suspend fun downloadAllServerChanges(timeout: Duration): Boolean {
+        return waitForChanges(TransferDirection.DOWNLOAD, timeout)
+    }
+
+    override suspend fun uploadAllLocalChanges(timeout: Duration): Boolean {
+        return waitForChanges(TransferDirection.UPLOAD, timeout)
+    }
+
+    /**
+     * Wrap Core callbacks that will not be invoked until data has been either fully uploaded
+     * or downloaded.
+     *
+     * When this method returns. The user facing Realm has been updated to the latest state.
+     *
+     * @param direction whether data is being uploaded or downloaded.
+     * @param timeout timeout parameter.
+     * @return `true` if the job completed before the timeout was hit, `false` otherwise.
+     */
+    private suspend fun waitForChanges(direction: TransferDirection, timeout: Duration): Boolean {
+        if (realm == null) {
```

Comment:
Is this the only way we assume we're in a `SyncSession.ErrorHandler`?

Reply:
Yes, it is a bit of a hack. But currently the only place we construct a session without a Realm is inside the error handler.  I can add a comment here to make that more clear?